subarray.go

```diff
@@ -0,0 +1,266 @@
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error setting Config: %s", sa.context.LastError())
+	}
+	return nil
+}
+
+// Free releases the internal TileDB core data that was allocated on the C heap.
+// It is automatically called when this object is garbage collected, but can be called earlier to
+// manually release memory if needed. Free is idempotent and can safely be called many times on the same object.
+func (sa *Subarray) Free() {
+	if sa.subarray != nil {
+		C.tiledb_subarray_free(&sa.subarray)
+	}
+}
+
+// SetCoalesceRanges sets coalesce_ranges property on a TileDB subarray object.
+// Intended to be used just after array.NewSubarray to replace the initial coalesce_ranges == true with coalesce_ranges = false if needed.
+func (sa *Subarray) SetCoalesceRanges(b bool) error {
+	var coalesce C.int
+	if b {
+		coalesce = 1
+	}
+
+	ret := C.tiledb_subarray_set_coalesce_ranges(sa.context.tiledbContext, sa.subarray, coalesce)
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error setting coalesce ranges on subarray: %s", sa.context.LastError())
+	}
+
+	return nil
+}
+
+// AddRange adds a range along a subarray dimension. It checks the types of range and dimension and
+// if the datatype of the range is not the same as the type of the dimension it returns an error.
+func (sa *Subarray) AddRange(dimIdx uint32, r Range) error {
+	dt, isVar, err := datatypeOfDimensionFromIndex(sa.array, dimIdx)
+	if err != nil {
+		return err
+	}
+	if err := r.assertCompatibility(dt, isVar); err != nil {
+		return err
+	}
+
+	var ret C.int32_t
+	var startSlice, endSlice []byte
+	if isVar {
+		startSlice := []byte(r.start.(string))
+		endSlice := []byte(r.end.(string))
+		ret = C.tiledb_subarray_add_range_var(sa.context.tiledbContext, sa.subarray, C.uint32_t(dimIdx),
+			unsafe.Pointer(&startSlice[0]), C.uint64_t(len(startSlice)), unsafe.Pointer(&endSlice[0]), C.uint64_t(len(endSlice)))
+	} else {
+		ret = C.tiledb_subarray_add_range(sa.context.tiledbContext, sa.subarray, C.uint32_t(dimIdx),
+			addressableValue(r.start).UnsafePointer(), addressableValue(r.end).UnsafePointer(), nil)
+	}
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error adding subarray range: %s", sa.context.LastError())
+	}
+
+	runtime.KeepAlive(startSlice)
+	runtime.KeepAlive(endSlice)
+	runtime.KeepAlive(r)
+
+	return nil
+}
+
+// AddRangeByName adds a range along a subarray dimension. It checks the types of range and dimension and
+// if the datatype of the range is not the same as the type of the dimension it returns an error.
+func (sa *Subarray) AddRangeByName(dimName string, r Range) error {
+	dt, isVar, err := datatypeOfDimensionFromName(sa.array, dimName)
+	if err != nil {
+		return err
+	}
+	if err := r.assertCompatibility(dt, isVar); err != nil {
+		return err
+	}
+
+	cDimName := C.CString(dimName)
+	defer C.free(unsafe.Pointer(cDimName))
+
+	var ret C.int32_t
+	var startSlice, endSlice []byte
```

Comment:
these can move into `if isVar` and become `startSlice := []byte(...)`, and then the `runtime.KeepAlive(startSlice)` etc. can be at the end of the if statement.